flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/TableSourceJsonPlanTest_jsonplan/testPartitionPushDown.out

```diff
@@ -64,7 +61,7 @@
         "b" : "INT"
       } ]
     },
-    "description" : "TableSourceScan(table=[[default_catalog, default_database, PartitionTable, filter=[], partitions=[{p=A}], project=[a, b], metadata=[]]], fields=[a, b])",
+    "description" : "TableSourceScan(table=[[default_catalog, default_database, PartitionTable, partitions=[{p=A}], project=[a, b], metadata=[]]], fields=[a, b])",
```

Comment:
is the `toString` consistent now everywhere? first `partitions`, then `filter`, then `project`, then `metadata`. This is how a source should work.

Reply:
The specs will be print each by the order in which they added to the table source. Now PushDown Partitions will first and partitions info will be print first. I can add a test case about json plan to verify it when the filters and partitions are all exist in the table source.

Reply:
> specs will be print each by the order in which they added to the table source

My question if the print order is consistent everywhere for the specs. But I don't think we need more tests for now. Thanks.

Reply:
Sorry, I test some cases and find one bad case when the table source has a watermark. It will miss the rule in optimize period "predicate_pushdown" and match the "logical". So PushDownProject will be called first in "logical", and then PushDownWatermark will be called later in "logical_rewrite", and then PushDownFilter in "logical_rewrite". The source description just like :
`TableSourceScan(table=[[default_catalog, default_database, TestTable, project=[a, b, c], metadata=[], watermark=[c], filter=[>(a, 1)]]], fields=[a, b, c])`
Maybe we need to sort the specs in table source to print them in order if the order is important.

Reply:
Watermarks and filters are a tricky topic anyway, because if the filter is too good, you will have almost no progress in event-time anymore. Let's open a follow up issue for this. I think having a clear spec order could be beneficial.

Reply:
Got it! The issue is here [FLINK-24754].